src/file_storage/file_hdf5.c

```diff
@@ -0,0 +1,929 @@
+        free(data);
+        free(block);
+        free(stride);                
+        return dspaces_ERR_HDF5;
+    }
+
+    status = H5Sclose(memspace_id);
+    if(status < 0) {
+        fprintf(stderr, "HDF5 failed to close the memory space for data reading "
+                        "from dataset: %s in file: %s.\n", dataset_name, file_name);
+        free(data);
+        free(block);
+        free(stride); 
+        return dspaces_ERR_HDF5;
+    }
+
+    status = H5Sclose(dataspace_id);
+    if(status < 0) {
+        fprintf(stderr,"HDF5 failed to close the dataspace of dataset: %s in file: %s.\n", 
+                        dataset_name, file_name);
+        free(data);
+        free(block);
+        free(stride); 
+        return dspaces_ERR_HDF5;
+    }
+
+    status = H5Dclose(dataset_id);
+    if(status < 0) {
+        fprintf(stderr,"HDF5 failed to close the dataset: %s in file: %s.\n", 
+                        dataset_name, file_name);
+        free(data);
+        free(block);
+        free(stride); 
+        return dspaces_ERR_HDF5;
+    }
+
+    free(block);
+    free(stride);
+    return dspaces_SUCCESS;
+}
+
+static int hdf5_write_scalar_attr(const char* file_name, hid_t file_id,
+                                     const char *attr_name, hid_t attr_dtype, const void* value)
+{
+    hid_t dataspace_id, attr_id;
+    herr_t status;
+
+    dataspace_id = H5Screate(H5S_SCALAR);
+    if(dataspace_id == H5I_INVALID_HID) {
+        fprintf(stderr,"HDF5 failed to create a scalar dataspace for attribute: %s in file: %s.\n",
+                        attr_name, file_name);
+        return dspaces_ERR_HDF5;
+    }
+
+    attr_id = H5Acreate2(file_id, attr_name, attr_dtype, dataspace_id, H5P_DEFAULT, H5P_DEFAULT);
+    if(attr_id < 0) {
+        fprintf(stderr,"HDF5 failed to create attribute: %s in file: %s.\n", attr_name, file_name);
+        return dspaces_ERR_HDF5;
+    }
+
+    status = H5Awrite(attr_id, attr_dtype, value);
+    if(status < 0) {
+        fprintf(stderr,"HDF5 failed to write attribute: %s to file: %s.\n", attr_name, file_name);
+        return dspaces_ERR_HDF5;
+    }
+
+    status = H5Aclose(attr_id);
+    if(status < 0) {
+        fprintf(stderr,"HDF5 failed to close attribute: %s in file: %s.\n", attr_name, file_name);
+        return dspaces_ERR_HDF5;
+    }
+
+    status = H5Sclose(dataspace_id);
+    if(status < 0) {
+        fprintf(stderr,"HDF5 failed to close the scalar dataspace for attribute: %s in file: %s.\n",
+                        attr_name, file_name);
+        return dspaces_ERR_HDF5;
+    }
+
+    return dspaces_SUCCESS;
+}
+
+/* Write a dspaces obj_data to a HDF5 file.
+ * The HDF5 file name is "{odsc_name}.ver%u.h5"
+ * The HDF5 dataset name is the bbox formatted as
+ * "{lb[0],lb[1], ...}-{ub[0], ub[1], ...}".
+ * The written obj_data remains in dspaces memory after this write.
+ * */
+int hdf5_write_od(struct swap_config* swap_conf, struct obj_data *od)
```

Comment:
1. HDF5 has the default file lock preventing 2 processes open the file at the same time.
2. I'm not sure about the if it prevents 2 threads to do that, I might have to design a test and check it.

Reply:
The HDF5 doc has this: `HDF5_ENABLE_THREADSAFE=ON` that we can turn on when building HDF5. But I'm not sure how to force it if we are using other-built hdf5, like module on the HPC.

Reply:
Unfortunately, I think it requires a lock per file. Can you check if thread safety is enabled? If so, you could in theory do a compile-time check.

Reply:
lock-per-file is enabled as a default HDF5 building option. I think it is feasible to check the thread-safe marco, but I'd rather put this as a TODO. Given that we are now using mutex when swapping `od` out, I don't think this will make a chaos.

Reply:
I mean threadlock per file to prevent multiple threads from writing to the same file. You are right, it's moot if we are doing the whole swap in the critical section. If we decide we don't like the performance of that, we'll have to get more granular about locking.